Complete sprint 14

Evo-Eclipse opened the pull request:
Storyboard Removal & Programmatic UI Migration

- Removed Main.storyboard and LaunchScreen.storyboard; fully transitioned the app to programmatic UI
- Cleaned up Info.plist and project configuration from storyboard references
- Updated SceneDelegate to manually initialize the window and set SplashViewController as the root
- Moved AppDelegate, SceneDelegate, and Info.plist to the Application directory for improved structure
- Added splash screen (SplashViewController) with centered logo and custom background

Navigation & Tab Bar Improvements

- Introduced TabBarViewController with custom style and two tabs: Trackers and Statistics
- Updated SplashViewController to handle transition to the main app interface
- Wrapped TrackersViewController in UINavigationController for better navigation stack handling
- Added search bar and placeholder state to TrackersViewController
- Updated navigation bar: moved add button and date picker to navigationBar, improved positioning and custom images for buttons

Trackers Feature Expansion

- Implemented Trackers collection view using compositional layout for flexible presentation
- Created TrackerCell with card-based UI, emoji support, and completion button logic
- Added TrackerHeaderView for section headers by category
- Developed TrackerCellDelegate and corresponding logic for completion toggling by date (with proper handling for future dates)
- Empty state UI updated with placeholderStackView
- Added button completion assets for state indication

Tracker Creation Flow

- Implemented multi-screen modal flow for creating trackers:
    - TrackerTypeSelectionViewController, TrackerCreationFormViewController, TrackerCategorySelectionViewController, TrackerScheduleSelectionViewController
    - Delegate and closure patterns for passing data between screens
    - Added TrackerType enum for type/state management
    - Enhanced Weekday enum with computed UI properties
    - Integrated dynamic rounding on custom table cells
- TrackersViewController now acts as delegate to receive and display newly created trackers

UI Components & Code Refactoring

- Extracted CategoryCell, ScheduleCell, and SpacedTextField into the Views/ directory for reusability
- Simplified TrackerCategorySelectionViewController and TrackerScheduleSelectionViewController, removing duplicated UI code
- Improved code modularity and maintainability

Models & Data Handling

- Introduced Tracker, TrackerCategory, TrackerRecord models for structuring tracker logic and state
- Added WeekDay enum for scheduling support
- Integrated storage properties into TrackersViewController for managing tracker data

Small Enhancements & Configuration

- Assets: added color sets, splash logo, button and icon images to Resources/Assets.xcassets
- Project now supports only portrait orientation and a minimum deployment target of iOS 17

Evo-Eclipse committed:
Update Xcodeproj

Configures the project to support only portrait orientation.
Sets the minimum iOS deployment target to version 17.

Evo-Eclipse committed:
Implement splash screen and Remove storyboard dependencies

Storyboard removal:
- Removed Main.storyboard and LaunchScreen.storyboard, transitioning to fully programmatic UI
- Updated project configuration to remove storyboard references
- Cleaned up Info.plist by removing storyboard-related keys

Splash screen implementation:
- Introduced SplashViewController as the initial root view controller with centered logo and custom background
- Updated SceneDelegate to initialize window manually and set SplashViewController as root

Project structure:
- Moved AppDelegate, SceneDelegate, and Info.plist into Application directory
- Migrated ViewController into Controllers directory structure
- Updated project configuration to point to new Info.plist location

Assets and resources:
- Added YP color sets and splash logo assets to Resources/Assets.xcassets

Evo-Eclipse committed:
Add Tab Bar with Trackers/Statistics Controllers

- Created TabBarViewController with custom styling and tab configuration
- Added TrackersViewController with search bar and placeholder state
- Added StatisticsViewController
- Updated SplashViewController to transition to TabBarViewController
- Added button_add and icon_dizzy image assets

Evo-Eclipse committed:
Add Tracker Models

- Added Tracker, TrackerCategory, TrackerRecord models
- Added WeekDay enum for scheduling
- Added storage properties to TrackersViewController

Evo-Eclipse committed:
Add UINavigationController for Trackers & Refactor

Reorganized navigation structure and improved UI consistency:
- Wrapped TrackersViewController in UINavigationController in TabBarViewController
- Moved addButton and datePicker to navigation bar via setupNavigationBar
- Changed addButton type to .custom to match custom image usage
- Removed addButton and datePicker from view hierarchy and constraints; updated setupViews and setupConstraints accordingly
- Adjusted titleLabel positioning and constraints to reference safe area directly for clean layout

Evo-Eclipse committed:
Implement Trackers collection view

Added full collection view implementation with compositional layout, custom cells, and tracker management system

- Implemented UICollectionView with compositional layout for tracker display in TrackersViewController
- Created TrackerCell with card-based design, emoji display, and completion button functionality
- Added TrackerHeaderView for section headers with category titles
- Implemented UICollectionViewDataSource and UICollectionViewDelegate protocols
- Added tracker management methods (addTracker, removeTracker, updateUI, updatePlaceholderVisibility)
- Integrated placeholder visibility logic based on tracker availability
- Added button completion assets (button_complete, button_completed) for tracker states
- Refactored SplashViewController setupViews method for cleaner code structure

Evo-Eclipse committed:
Implement Tracker creation flow

Added multi-screen modal flow for creating new trackers with type selection, form input, and scheduling.

- Added TrackerTypeSelectionViewController, TrackerCreationFormViewController, TrackerCategorySelectionViewController, TrackerScheduleSelectionViewController
- Updated TrackersViewController with NewTrackerViewControllerDelegate for receiving new trackers
- Enhanced Weekday enum with UI computed properties
- Implemented delegate pattern and closures for data passing between screens
- Added custom table cells (CategoryCell, ScheduleCell) with dynamic corner rounding
- Created TrackerType enum for state management and UI variations

Flow: Request → TrackerTypeSelection → TrackerCreationForm ←→ (Category/Schedule selection) → Return

Evo-Eclipse committed:
Refactor reusable UI components

Extracted custom table cells and text field to dedicated files, improving modularity and code organization

- Moved CategoryCell, ScheduleCell, and SpacedTextField from controllers to Views/ as standalone classes
- Simplified TrackerCategorySelectionViewController and TrackerScheduleSelectionViewController by referencing new view components
- Removed duplicated cell and text field code from respective controllers

Evo-Eclipse committed:
Implement Tracker counter & completion toggle

- Added TrackerCellDelegate and implemented completion state/counter logic in TrackersViewController and TrackerCell
- Introduced Int extension with pluralization for day counter display
- Updated empty state UI: replaced placeholder views with placeholderStackView
- Complete button now updates state per date, disables for future dates, and refreshes instantly on toggle

IgorChumikov commented on Tracker/Controllers/TrackersViewController.swift:
```diff
@@ -0,0 +1,368 @@
+        searchBar.resignFirstResponder()
+    }
+}
+
```
Пунктом 6.5:
🔍 Что требует пункт:
При изменении даты отображаются трекеры привычек, которые должны быть видны в день недели, выбранный в UIDatePicker.

Это означает:
У каждого трекера есть расписание ([Weekday]);
Когда пользователь выбирает дату в UIDatePicker, надо:
1 получить день недели этой даты (например, среда);
2 отфильтровать все трекеры и показать только те, у которых schedule содержит этот день;
3 также показать все нерегулярные события, у которых schedule пустой или hasSchedule == false.

❌ Что не так в текущей реализации:
1. Фильтрация трекеров по Weekday не реализована вообще
В TrackersViewController этого кода нет:
```

let selectedWeekday = Weekday.from(date: currentDate)
filteredTrackers = allTrackers.filter { $0.schedule.contains(selectedWeekday) }

```

Твоё текущее обновление в @objc private func dateChanged(_ sender: UIDatePicker):
```
print("Date changed to: \(formatter.string(from: sender.date))")
collectionView.reloadData()
```

➡️ просто вызывает reloadData() без какой-либо фильтрации.

2. Массив categories используется напрямую без учёта выбранной даты
```
func numberOfSections(in collectionView: UICollectionView) -> Int {
    return categories.count
}
```
и далее:
```
categories[indexPath.section].trackers[indexPath.item]
```
➡️ здесь всегда берутся все трекеры, без фильтрации по текущему дню.

✅ Как должно быть
Чтобы исправить пункт 6.5, нужно:

✅ 1. Добавить currentDate
```
private var currentDate: Date = Date()

```
И в dateChanged — обновлять это значение:
```
@objc private func dateChanged(_ sender: UIDatePicker) {
    currentDate = sender.date
    filterVisibleTrackers()
}
```

✅ 2. Добавить массив visibleCategories
```
private var visibleCategories: [TrackerCategory] = []
```

✅ 3. Реализовать метод фильтрации:
```
private func filterVisibleTrackers() {
    let calendar = Calendar.current
    let weekdayIndex = calendar.component(.weekday, from: currentDate) // Sunday = 1 ... Saturday = 7
    let weekday = Weekday.fromSystemIndex(weekdayIndex)
    
    visibleCategories = categories.compactMap { category in
        let filtered = category.trackers.filter {
            !$0.schedule.isEmpty && $0.schedule.contains(weekday)
        }
        return filtered.isEmpty ? nil : TrackerCategory(title: category.title, trackers: filtered)
    }
    
    collectionView.reloadData()
    updatePlaceholderVisibility()
}
```

✅ 4. Использовать visibleCategories вместо categories:
```
func numberOfSections(in collectionView: UICollectionView) -> Int {
    return visibleCategories.count
}

func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
    return visibleCategories[section].trackers.count
}

func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
    let tracker = visibleCategories[indexPath.section].trackers[indexPath.item]
    ...
}
```

✅ 5. Добавить утилиту для преобразования system weekday → твой Weekday
```
extension Weekday {
    static func fromSystemIndex(_ index: Int) -> Weekday {
        // Sunday = 1, Monday = 2, ..., Saturday = 7
        // У нас monday = 0
        switch index {
        case 1: return .sunday
        case 2: return .monday
        case 3: return .tuesday
        case 4: return .wednesday
        case 5: return .thursday
        case 6: return .friday
        case 7: return .saturday
        default: return .monday
        }
    }
}
```

IgorChumikov commented on Tracker/Controllers/TrackersViewController.swift:
```diff
@@ -0,0 +1,368 @@
+        }
+    }
+}
+
```
Пункт 8.13

Что в текущем коде:
В твоём TrackersViewController нет такого свойства:
```
private var currentDate: Date = Date()
```
Ты используешь datePicker.date напрямую:
```

@objc private func dateChanged(_ sender: UIDatePicker) {
    print("Date changed to: \(formatter.string(from: sender.date))")
    collectionView.reloadData()
}

```

И при создании трекера:
```
cell.configure(with: tracker, on: datePicker.date)
```

Почему это плохо
- Невозможно отслеживать текущее значение даты без прямого доступа к UIDatePicker — нарушает инкапсуляцию;
- Если завтра ты уберёшь UIDatePicker и будешь использовать другой источник, код сломается;
- Фильтрация по расписанию (из пункта 6.5) должна использовать именно currentDate, как источник "активного дня".

Как должно быть:
1. Добавь свойство:
```
private var currentDate: Date = Date()
```

2. Обновляй его при изменении:
```
@objc private func dateChanged(_ sender: UIDatePicker) {
    currentDate = sender.date
    filterVisibleTrackers()
}
```
3. Используй его в:
 - filterVisibleTrackers() для определения дня недели;
 - TrackerCell.configure(with:tracker,on:currentDate).
 
 
 🛠 Что нужно сделать, чтобы ✅:

1 В TrackersViewController:
```
private var currentDate: Date = Date()
```
2 В dateChanged(_:):
```
currentDate = sender.date
filterVisibleTrackers()
```
3 В TrackerCell.configure(...):
```
cell.configure(with: tracker, on: currentDate)
```

IgorChumikov commented on Tracker/Controllers/TrackersViewController.swift:
```diff
@@ -0,0 +1,368 @@
+}
+
+// MARK: - Tracker Management
+
```
8.14 
 
 Что не так в твоём проекте:
1. Свойства currentDate в TrackersViewController нет
(об этом мы уже говорили в пункте 8.13)

2. Дата из UIDatePicker не используется для фильтрации трекеров по дню недели
Вот твой текущий обработчик выбора даты:
```

@objc private func dateChanged(_ sender: UIDatePicker) {
    print("Date changed to: \(formatter.string(from: sender.date))")
    collectionView.reloadData()
}

```
🔴 Здесь нет логики фильтрации по расписанию.

3. collectionView(_:numberOfItemsInSection:) и cellForItemAt используют categories, не visibleCategories или отфильтрованный список.

✅ Как должно быть реализовано:

1. Добавить currentDate:
```
private var currentDate: Date = Date()
```
2. В @objc private func dateChanged(_:):
```
@objc private func dateChanged(_ sender: UIDatePicker) {
    currentDate = sender.date
    filterVisibleTrackers()
}
```

3. Метод filterVisibleTrackers():
```

private func filterVisibleTrackers() {
    let calendar = Calendar.current
    let weekdayIndex = calendar.component(.weekday, from: currentDate) // Sunday = 1
    let weekday = Weekday.fromSystemIndex(weekdayIndex)

    visibleCategories = categories.compactMap { category in
        let filteredTrackers = category.trackers.filter {
            $0.schedule.contains(weekday)
        }

        return filteredTrackers.isEmpty ? nil : TrackerCategory(title: category.title, trackers: filteredTrackers)
    }

    collectionView.reloadData()
}

```


4. Использовать visibleCategories в dataSource:
```
func numberOfSections(in collectionView: UICollectionView) -> Int {
    return visibleCategories.count
}
func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
    return visibleCategories[section].trackers.count
}

```

📎 Дополнительно: Weekday.fromSystemIndex(_:)
```

extension Weekday {
    static func fromSystemIndex(_ index: Int) -> Weekday {
        switch index {
        case 1: return .sunday
        case 2: return .monday
        case 3: return .tuesday
        case 4: return .wednesday
        case 5: return .thursday
        case 6: return .friday
        case 7: return .saturday
        default: return .monday
        }
    }
}

```


💡 Вывод:
Пункт 8.14 не выполнен, потому что:
нет currentDate;
нет фильтрации по дню недели;
dateChanged(_:) не запускает фильтрацию.

IgorChumikov commented on Tracker/Controllers/TrackersViewController.swift:
```diff
@@ -0,0 +1,368 @@
+            self?.updatePlaceholderVisibility()
+        }
+    }
+    
```
пункт 9.7
 
 🛠 Что нужно исправить:
Добавить "По умолчанию" в массив categories;
В TrackerCreationFormViewController — подставлять её, если selectedCategory == nil.

Evo-Eclipse committed:
Implement date-based Tracker filtering & default category

Added tracker filtering by selected date and introduced default category fallback for improved user experience

Weekday.swift:

- Refactored Weekday enum into separate file with fromSystemIndex() conversion method

TrackersViewController.swift:

- Added currentDate property and visibleCategories array for date-based filtering
- Implemented filterVisibleTrackers() method with weekday matching logic
- Reorganized methods by moving tracker management inline
- Updated delegate conformance to TrackerCreationFormViewControllerDelegate

TrackerCreationFormViewController.swift:

- Renamed protocol from NewTrackerViewControllerDelegate to TrackerCreationFormViewControllerDelegate
- Fixed header comment to match actual class name
- Introduced default category as fallback for uncategorized trackers
- Updated form validation to remove mandatory category requirement

TrackerCategorySelectionViewController.swift:

- Introduced default category as fallback for uncategorized trackers
- Updated form validation to remove mandatory category requirement

TrackerTypeSelectionViewController.swift:

- Updated delegate property to use renamed TrackerCreationFormViewControllerDelegate

Navigation Controllers:

- Added navigationItem.hidesBackButton = true across all creation flow controllers
- Applied to TrackerTypeSelection, TrackerCreationForm, TrackerScheduleSelection, TrackerCategorySelection

Evo-Eclipse committed:
Enhance keyboard dismiss UX

Introduced dismissKeyboardOnTap extension for UIViewController to improve keyboard handling on form and category screens

UIViewController+dismissKeyboardOnTap.swift:

- Added global dismissKeyboardOnTap() extension with tap gesture integration for keyboard hiding
- Provided private @objc method to end editing when user taps outside input field

TrackerCategorySelectionViewController.swift, TrackerCreationFormViewController.swift:

- Enabled dismissKeyboardOnTap() across creation and category selection flows for seamless input experience

TrackerCreationFormViewController.swift:

- Moved view.backgroundColor assignment to viewDidLoad for clarity
- Replaced UITextFieldDelegate from textFieldDidChangeSelection to textFieldShouldReturn for return key handling and resigning first responder
